snowfakery/data_generator_runtime.py

```diff
@@ -119,6 +121,45 @@ class Dependency(NamedTuple):
 )
 
 
+class SlotState(Enum):
+    UNUSED = auto()
+    ALLOCATED = auto()
+    CONSUMED = auto()
+
+
+class NicknameSlot:
+    _tablename: str
+    id_manager: IdManager
+    allocated_id: int = None
+
+    def __init__(self, tablename, id_manager):
+        self._tablename = tablename
+        self.id_manager = id_manager
+
+    @property
+    def id(self):
+        if self.allocated_id is None:
+            self.allocated_id = self.id_manager.generate_id(self._tablename)
+        return self.allocated_id
+
+    def consume_slot(self):
+        rc = self.allocated_id
+        self.allocated_id = SlotState.CONSUMED
+        return rc
+
+    @property
+    def status(self):
+        if self.allocated_id is None:
+            return SlotState.UNUSED
+        elif isinstance(self.allocated_id, int):
+            return SlotState.ALLOCATED
+        elif self.allocated_id == SlotState.CONSUMED:
+            return SlotState.CONSUMED
```

Comment:
Isn't there a risk of collision between the SlotState enum integers and actual object ids?

Reply:
No, the enums are pretty robust from a type perspective.

```
>>> SlotState.UNUSED
<SlotState.UNUSED: 1>
>>> SlotState.UNUSED==1
False
>>> int(SlotState.UNUSED)==1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: int() argument must be a string, a bytes-like object or a number, not 'SlotState'
>>> SlotState.UNUSED.value == 1
True
```